Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -76,6 +110,107 @@ TextureMapperLayer::~TextureMapperLayer()
     removeFromParent();
 }
 
+void TextureMapperLayer::flattenDescendantLayersIfNecessary(TextureMapperPaintOptions& options)
+{
+    // Preprocess is called as a first step before painting any layers
+
+    // Layers are collected in depth first, post-order
+    Vector<TextureMapperLayer*> layersToFlatten;
+    collectLayersToFlattenRecursive(layersToFlatten);
+
+    for (auto* layer : layersToFlatten)
+        layer->flatten(options);
+}
+
+void TextureMapperLayer::collectLayersToFlattenRecursive(Vector<TextureMapperLayer*>& layers)
+{
+    // Traverse all children
+    for (auto* child : m_children) {
+        // Recursively collect from the child's descendants
+        child->collectLayersToFlattenRecursive(layers);
+
+        // If the child satisfies the condition, add it to the layers vector
+        if (child->flattensAsLeafOf3DSceneOr3DPerspective())
+            layers.append(child);
+    }
+}
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
+{
+    // Flattens descendant layers to into this planes z=0. In theory we could
+    // project each sublayer to this layer's plane but in practice that doesn't
+    // result in pixel perfect outcome and that's why each sublayer needs to be
+    // rendered into this layer's in 2D root.
+
+    // Reset transformations for flattening root 2D
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
```

Comment:
Windows port is broken.
https://gist.github.com/fujii/dd5e25cbc0f5241a360390033c4fed84

Reply:
Thanks, I applied the change